File: gluecodium/src/test/resources/smoke/defaults_const/output/swift/smoke/EnumCollectionDefaults.swift
//
//
import Foundation
public struct EnumCollectionDefaults {
    public var listField: [Enum1]
    public var setField: Set<Enum2>
    public var mapField: [Enum3: Enum4]
    public init(listField: [Enum1] = [Enum1.disabled], setField: Set<Enum2> = [Enum2.disabled], mapField: [Enum3: Enum4] = [Enum3.disabled: Enum4.disabled]) {
        self.listField = listField
        self.setField = setField
        self.mapField = mapField
    }
    internal init(cHandle: _baseRef) {
        listField = moveFromCType(smoke_EnumCollectionDefaults_listField_get(cHandle))
        setField = moveFromCType(smoke_EnumCollectionDefaults_setField_get(cHandle))
        mapField = moveFromCType(smoke_EnumCollectionDefaults_mapField_get(cHandle))
    }
}
internal func copyFromCType(_ handle: _baseRef) -> EnumCollectionDefaults {
    return EnumCollectionDefaults(cHandle: handle)
}
internal func moveFromCType(_ handle: _baseRef) -> EnumCollectionDefaults {
    defer {
        smoke_EnumCollectionDefaults_release_handle(handle)
    }
    return copyFromCType(handle)
}
internal func copyToCType(_ swiftType: EnumCollectionDefaults) -> RefHolder {
    let c_listField = moveToCType(swiftType.listField)
    let c_setField = moveToCType(swiftType.setField)
    let c_mapField = moveToCType(swiftType.mapField)
    return RefHolder(smoke_EnumCollectionDefaults_create_handle(c_listField.ref, c_setField.ref, c_mapField.ref))
}
internal func moveToCType(_ swiftType: EnumCollectionDefaults) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_EnumCollectionDefaults_release_handle)
}
internal func copyFromCType(_ handle: _baseRef) -> EnumCollectionDefaults? {
    guard handle != 0 else {
        return nil
    }
    let unwrappedHandle = smoke_EnumCollectionDefaults_unwrap_optional_handle(handle)
    return EnumCollectionDefaults(cHandle: unwrappedHandle) as EnumCollectionDefaults
}
internal func moveFromCType(_ handle: _baseRef) -> EnumCollectionDefaults? {
    defer {
        smoke_EnumCollectionDefaults_release_optional_handle(handle)
    }
    return copyFromCType(handle)
}
internal func copyToCType(_ swiftType: EnumCollectionDefaults?) -> RefHolder {
    guard let swiftType = swiftType else {
        return RefHolder(0)
    }
    let c_listField = moveToCType(swiftType.listField)
    let c_setField = moveToCType(swiftType.setField)
    let c_mapField = moveToCType(swiftType.mapField)
    return RefHolder(smoke_EnumCollectionDefaults_create_optional_handle(c_listField.ref, c_setField.ref, c_mapField.ref))
}
internal func moveToCType(_ swiftType: EnumCollectionDefaults?) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_EnumCollectionDefaults_release_optional_handle)
}

File: gluecodium/src/test/resources/smoke/defaults_const/output/dart/lib/src/smoke/enum_collection_defaults.dart
import 'dart:ffi';
import 'package:library/src/_library_context.dart' as __lib;
import 'package:library/src/fire/enum1.dart';
import 'package:library/src/fire/enum2.dart';
import 'package:library/src/fire/enum3.dart';
import 'package:library/src/fire/enum4.dart';
import 'package:library/src/generic_types__conversion.dart';
class EnumCollectionDefaults {
  List<Enum1> listField;
  Set<Enum2> setField;
  Map<Enum3, Enum4> mapField;
  EnumCollectionDefaults._(this.listField, this.setField, this.mapField);
  EnumCollectionDefaults()
    : listField = [Enum1.disabled], setField = {Enum2.disabled}, mapField = {Enum3.disabled: Enum4.disabled};
}
// EnumCollectionDefaults "private" section, not exported.
final _smokeEnumcollectiondefaultsCreateHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>, Pointer<Void>, Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>, Pointer<Void>, Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_create_handle'));
final _smokeEnumcollectiondefaultsReleaseHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_release_handle'));
final _smokeEnumcollectiondefaultsGetFieldlistField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_get_field_listField'));
final _smokeEnumcollectiondefaultsGetFieldsetField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_get_field_setField'));
final _smokeEnumcollectiondefaultsGetFieldmapField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_get_field_mapField'));
Pointer<Void> smokeEnumcollectiondefaultsToFfi(EnumCollectionDefaults value) {
  final _listFieldHandle = listofFireEnum1ToFfi(value.listField);
  final _setFieldHandle = setofFireEnum2ToFfi(value.setField);
  final _mapFieldHandle = mapofFireEnum3ToFireEnum4ToFfi(value.mapField);
  final _result = _smokeEnumcollectiondefaultsCreateHandle(_listFieldHandle, _setFieldHandle, _mapFieldHandle);
  listofFireEnum1ReleaseFfiHandle(_listFieldHandle);
  setofFireEnum2ReleaseFfiHandle(_setFieldHandle);
  mapofFireEnum3ToFireEnum4ReleaseFfiHandle(_mapFieldHandle);
  return _result;
}
EnumCollectionDefaults smokeEnumcollectiondefaultsFromFfi(Pointer<Void> handle) {
  final _listFieldHandle = _smokeEnumcollectiondefaultsGetFieldlistField(handle);
  final _setFieldHandle = _smokeEnumcollectiondefaultsGetFieldsetField(handle);
  final _mapFieldHandle = _smokeEnumcollectiondefaultsGetFieldmapField(handle);
  try {
    return EnumCollectionDefaults._(
      listofFireEnum1FromFfi(_listFieldHandle),
      setofFireEnum2FromFfi(_setFieldHandle),
      mapofFireEnum3ToFireEnum4FromFfi(_mapFieldHandle)
    );
  } finally {
    listofFireEnum1ReleaseFfiHandle(_listFieldHandle);
    setofFireEnum2ReleaseFfiHandle(_setFieldHandle);
    mapofFireEnum3ToFireEnum4ReleaseFfiHandle(_mapFieldHandle);
  }
}
void smokeEnumcollectiondefaultsReleaseFfiHandle(Pointer<Void> handle) => _smokeEnumcollectiondefaultsReleaseHandle(handle);
// Nullable EnumCollectionDefaults
final _smokeEnumcollectiondefaultsCreateHandleNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_create_handle_nullable'));
final _smokeEnumcollectiondefaultsReleaseHandleNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_release_handle_nullable'));
final _smokeEnumcollectiondefaultsGetValueNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_EnumCollectionDefaults_get_value_nullable'));
Pointer<Void> smokeEnumcollectiondefaultsToFfiNullable(EnumCollectionDefaults? value) {
  if (value == null) return Pointer<Void>.fromAddress(0);
  final _handle = smokeEnumcollectiondefaultsToFfi(value);
  final result = _smokeEnumcollectiondefaultsCreateHandleNullable(_handle);
  smokeEnumcollectiondefaultsReleaseFfiHandle(_handle);
  return result;
}
EnumCollectionDefaults? smokeEnumcollectiondefaultsFromFfiNullable(Pointer<Void> handle) {
  if (handle.address == 0) return null;
  final _handle = _smokeEnumcollectiondefaultsGetValueNullable(handle);
  final result = smokeEnumcollectiondefaultsFromFfi(_handle);
  smokeEnumcollectiondefaultsReleaseFfiHandle(_handle);
  return result;
}
void smokeEnumcollectiondefaultsReleaseFfiHandleNullable(Pointer<Void> handle) =>
  _smokeEnumcollectiondefaultsReleaseHandleNullable(handle);
// End of EnumCollectionDefaults "private" section.
